File: simulation/hardware/protocol.js
/**
 * protocol.js
 * STM32 Serial & WebSocket Communication Protocol Specifications.
 * 
 * Defines standard newline-delimited JSON commands and telemetry schemas:
 * 
 * Host -> STM32:
 *   {"cmd":"MOVE","pan":1200,"tilt":-350}
 *   {"cmd":"STOP"}
 *   {"cmd":"HOME"}
 *   {"cmd":"ENABLE","value":1}
 *   {"cmd":"ENABLE","value":0}
 * 
 * STM32 -> Host Telemetry:
 *   {
 *     "type":"telemetry",
 *     "pan":1187,
 *     "tilt":-342,
 *     "pan_target":1200,
 *     "tilt_target":-350,
 *     "pan_temp":31.4,
 *     "tilt_temp":29.8,
 *     "heater":0
 *   }
 * 
 * Status:
 *   {"type":"status","connected":true}
 */

export const PROTOCOL_CMDS = {
  MOVE: 'MOVE',
  STOP: 'STOP',
  HOME: 'HOME',
  ENABLE: 'ENABLE'
};

export const PROTOCOL_TYPES = {
  TELEMETRY: 'telemetry',
  STATUS: 'status',
  ACK: 'ack',
  ERROR: 'error'
};

/**
 * Build a MOVE command object with position and optical pixel error metadata
 * @param {number} pan Absolute pan steps
 * @param {number} tilt Absolute tilt steps
 * @param {Object} meta Metadata including errorPanPx, errorTiltPx, hasTarget
 * @returns {Object}
 */
export function buildMoveCmd(pan, tilt, meta = {}) {
  return {
    cmd: PROTOCOL_CMDS.MOVE,
    pan: Math.round(pan),
    tilt: Math.round(tilt),
    err_p: meta.errorPanPx !== undefined ? Math.round(meta.errorPanPx) : 0,
    err_t: meta.errorTiltPx !== undefined ? Math.round(meta.errorTiltPx) : 0,
    lock: meta.hasTarget !== undefined ? Boolean(meta.hasTarget) : true
  };
}

/**
 * Build a STOP command object
 * @returns {Object}
 */
export function buildStopCmd() {
  return { cmd: PROTOCOL_CMDS.STOP };
}

/**
 * Build a HOME command object
 * @returns {Object}
 */
export function buildHomeCmd() {
  return { cmd: PROTOCOL_CMDS.HOME };
}

/**
 * Build an ENABLE/DISABLE command object
 * @param {boolean|number} enable
 * @returns {Object}
 */
export function buildEnableCmd(enable) {
  return {
    cmd: PROTOCOL_CMDS.ENABLE,
    value: enable ? 1 : 0
  };
}

/**
 * Encode object to newline-delimited JSON string for serial / websocket
 * @param {Object} obj 
 * @returns {string}
 */
export function encodePacket(obj) {
  return JSON.stringify(obj) + '\n';
}

/**
 * Decode raw string into JSON object safely
 * @param {string} raw 
 * @returns {Object|null}
 */
export function decodePacket(raw) {
  if (!raw || typeof raw !== 'string') return null;
  const trimmed = raw.trim();
  if (!trimmed.startsWith('{') || !trimmed.endsWith('}')) return null;
  try {
    return JSON.parse(trimmed);
  } catch (err) {
    return null;
  }
}

/**
 * Validate incoming telemetry packet
 * @param {Object} pkt 
 * @returns {boolean}
 */
export function isValidTelemetry(pkt) {
  return (
    pkt &&
    pkt.type === PROTOCOL_TYPES.TELEMETRY &&
    typeof pkt.pan === 'number' &&
    typeof pkt.tilt === 'number'
  );
}

File: simulation/bridge/protocol.py
"""
protocol.py
Python Protocol Parser and Packet Schema for STM32 Hardware Bridge.
"""

import json

CMD_MOVE = "MOVE"
CMD_STOP = "STOP"
CMD_HOME = "HOME"
CMD_ENABLE = "ENABLE"

TYPE_TELEMETRY = "telemetry"
TYPE_STATUS = "status"

def encode_packet(data: dict) -> bytes:
    """Encode dictionary into newline-delimited JSON bytes."""
    return (json.dumps(data) + "\n").encode("utf-8")

def decode_packet(line: str) -> dict | None:
    """Decode raw string into JSON dictionary safely."""
    if not line:
        return None
    trimmed = line.strip()
    if not (trimmed.startswith("{") and trimmed.endswith("}")):
        return None
    try:
        return json.loads(trimmed)
    except Exception:
        return None

def build_telemetry(pan: int, tilt: int, pan_target: int, tilt_target: int,
                    pan_temp: float = 24.5, tilt_temp: float = 23.2, heater: int = 0) -> dict:
    """Build standard telemetry payload."""
    return {
        "type": TYPE_TELEMETRY,
        "pan": pan,
        "tilt": tilt,
        "pan_target": pan_target,
        "tilt_target": tilt_target,
        "pan_temp": round(pan_temp, 1),
        "tilt_temp": round(tilt_temp, 1),
        "heater": heater
    }
